Mathlib/Data/ENNReal/Floor.lean

```diff
@@ -0,0 +1,162 @@
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kalle Kytölä
+-/
+import Mathlib.Topology.Instances.ENNReal
+import Mathlib.Data.Real.ENatENNReal
+
+/-!
+# A floor function on the extended non-negative reals
+-/
+
+open Filter BigOperators TopologicalSpace Topology Set ENNReal NNReal ENat
+
+section ENat_topology
+
+instance : TopologicalSpace ℕ∞ := TopologicalSpace.induced ENat.toENNReal inferInstance
+
+lemma ENat.continuous_toENNReal : Continuous ENat.toENNReal :=
+  continuous_iff_le_induced.mpr fun _ h ↦ h
+
+instance : OrderTopology ℕ∞ := sorry
+
+-- TODO: Move to the appropriate file `Data.ENat.Basic`.
+lemma ENat.toENNReal_coe_eq_top_iff {m : ℕ∞} :
+    (m : ℝ≥0∞) = ∞ ↔ m = ⊤ := by
+  rw [← toENNReal_coe_eq_iff, toENNReal_top]
+
+-- TODO: Move to the appropriate file `Data.ENat.Basic`.
+lemma ENat.toENNReal_coe_ne_top_iff {m : ℕ∞} :
+    (m : ℝ≥0∞) ≠ ∞ ↔ m ≠ ⊤ :=
+  not_iff_not.mpr toENNReal_coe_eq_top_iff
+
+end ENat_topology
+
+section ENNReal_floor
+
+namespace ENNReal
+
+/-- The floor function `ℝ≥0∞ → ℕ∞`: the floor of `x` is the supremum of the extended natural
+numbers `n` satisfying `n ≤ x`. -/
+noncomputable def floor (x : ℝ≥0∞) : ℕ∞ := sSup {n : ℕ∞ | n ≤ x}
+
+variable {x y : ℝ≥0∞} {n : ℕ} {m : ℕ∞}
+
+lemma floor_eq_sSup_range_toENNReal_inter_Iic :
+    floor x = sSup (Set.range (fun (m : ℕ∞) ↦ (m : ℝ≥0∞)) ∩ Iic x) := by
+  simp only [floor]
+  rw [toENNReal_mono.map_sSup_of_continuousAt ENat.continuous_toENNReal.continuousAt (by simp)]
+  congr
+  ext m
+  simp only [mem_image, mem_setOf_eq, mem_inter_iff, mem_range, mem_Iic]
+  refine ⟨fun hx ↦ ?_, fun hx ↦ ?_⟩
+  · obtain ⟨n, hn⟩ := hx
+    refine ⟨⟨n, hn.2⟩, hn.2 ▸ hn.1⟩
+  · obtain ⟨n, hn⟩ := hx.1
+    refine ⟨n, ⟨hn ▸ hx.2, hn⟩⟩
+
+lemma floor_mono (h : x ≤ y) : x.floor ≤ y.floor := sSup_le_sSup <| fun _ hx ↦ hx.trans h
+
+/-- The floor function `ℝ≥0∞ → ℝ≥0∞` is increasing. -/
+lemma monotone_floor : Monotone floor := fun _ _ h ↦ floor_mono h
+
+lemma floor_le (x : ℝ≥0∞) : floor x ≤ x := by
+  simpa only [floor_eq_sSup_range_toENNReal_inter_Iic] using sSup_le fun _ h ↦ h.2
+
+lemma le_floor (h : m ≤ x) : m ≤ x.floor := le_sSup <| by simp [h]
+
+lemma floor_le_of_lt_add_one (h : x < m + 1) :
+    x.floor ≤ m := by
+  cases' m with l
+  · exact le_top
+  · apply sSup_le
+    intro n hn
+    have obs := lt_of_le_of_lt hn h
+    cases' n with k
+    · exfalso; apply not_top_lt obs
+    · exact Nat.cast_le.mpr <| Nat.le_of_lt_succ (by exact_mod_cast obs)
+
+@[simp] lemma floor_coe_eq (n : ℕ∞) :
+    floor n = n :=
+  le_antisymm (by exact_mod_cast floor_le n) (le_sSup (by simp))
+
+@[simp] lemma floor_zero_eq : floor 0 = 0 := by exact_mod_cast ENNReal.floor_coe_eq 0
+
+@[simp] lemma floor_one_eq_one : floor 1 = 1 := by exact_mod_cast ENNReal.floor_coe_eq 1
+
+@[simp] lemma floor_top_eq : floor ∞ = ⊤ := floor_coe_eq ⊤
+
+lemma floor_eq_natFloor_toNNReal (x_ne_top : x ≠ ∞) :
+    floor x = Nat.floor x.toNNReal := by
```

Comment:
Maybe worth turning around and simping

Reply:
This I am slightly unsure of, but I could test. My doubt is because `ENNReal`s can be complicated to work with, and to prove things about them, one still very often breaks to cases and handles the finite case by casting to `NNReal`. That is more or less how this lemma is used already below.